tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/PseudoLocales.java

```diff
@@ -0,0 +1,384 @@
+        }
+
+        private static ImmutableSet<String> getLocaleIds() {
+            return ID_MAP.keySet();
+        }
+
+        private final String localeId;
+        private final Function<Boolean, PseudoText> textSupplier;
+        // A string whose code points form the exemplar set for the pseudo locale.
+        private final String exemplars;
+
+        PseudoType(String localeId, Function<Boolean, PseudoText> textSupplier, String exemplars) {
+            this.localeId = localeId;
+            this.textSupplier = textSupplier;
+            this.exemplars = exemplars;
+        }
+
+        String getLocaleId() {
+            return localeId;
+        }
+
+        PseudoText getText(boolean isPattern) {
+            return textSupplier.apply(isPattern);
+        }
+
+        String getExemplars() {
+            return exemplars;
+        }
+    }
+
+    /**
+     * Returns a wrapped data supplier which will inject {@link CldrData} for the pseudo locales
+     * {@code en_XA} and {@code ar_XB}. These locales should behave in all respects like normal
+     * locales and can be processed accordingly.
+     */
+    public static CldrDataSupplier addPseudoLocalesTo(CldrDataSupplier src) {
+        return new PseudoSupplier(src);
+    }
+
+    private static final class PseudoSupplier extends CldrDataSupplier {
+        private final CldrDataSupplier src;
+        private final Set<String> srcIds;
+
+        PseudoSupplier(CldrDataSupplier src) {
+            this.src = checkNotNull(src);
+            this.srcIds = src.getAvailableLocaleIds();
+            PseudoType.getLocaleIds()
+                .forEach(id -> checkArgument(!srcIds.contains(id),
+                    "pseudo locale %s already supported", id));
+        }
+
+        @Override public CldrDataSupplier withDraftStatusAtLeast(CldrDraftStatus draftStatus) {
+            return new PseudoSupplier(src.withDraftStatusAtLeast(draftStatus));
+        }
+
+        @Override public CldrData getDataForLocale(String localeId, CldrResolution resolution) {
+            if (PseudoType.getLocaleIds().contains(localeId)) {
+                // Use resolved data to ensure we get all the values (e.g. values in "en_001").
+                CldrData enData = src.getDataForLocale("en", RESOLVED);
+                return new PseudoLocaleData(enData, resolution, PseudoType.fromId(localeId));
+            } else {
+                return src.getDataForLocale(localeId, resolution);
+            }
+        }
+
+        @Override public Set<String> getAvailableLocaleIds() {
+            return Sets.union(src.getAvailableLocaleIds(), PseudoType.getLocaleIds());
+        }
+
+        @Override public CldrData getDataForType(CldrDataType type) {
+            return src.getDataForType(type);
+        }
+    }
+
+    private interface PseudoText {
+        void addFragment(String text, boolean isLocalizable);
+    }
+
+    private static final class PseudoLocaleData extends FilteredData {
```

Comment:
Shouldn't the implementation of the pseudo-locale transformations remain in CLDR code, ideally with some public CLDR API? Or as an option on the CLDR data API??

ICU need not be the only user of pseudo-localized CLDR data.
For example, it should be possible to add pseudolocales to the CLDR JSON data.

Reply:
I'd be happy to move it once we agree on the API. I did it like this since it wasn't initially obvious how we would handle it in the common API (I actually initially didn't think the pseudo locales were legal CLDR language codes, but rather Google/ICU specific hacks).

Reply:
Ok, in that case please leave a TODO with CLDR & ICU follow-up ticket numbers.

FYI XA & XB officially listed in http://www.unicode.org/reports/tr35/tr35.html#unicode_region_subtag_validity

Reply:
Done. CLDR-13381 (not sure it also needs an ICU ticket though).